### PR TITLE
Add HTTP data binding support

### PR DESCRIPTION
**Pull request by chamil321**
## Automation tests
 - Unit tests 
   > Added
 - Integration tests
   > Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```
curl -v http://localhost:9090/hello/bindJson -X POST -d '{ "Details": { "ID": "77999", "Name": "XYZ"} , "Location": { "No": "01", "City": "Colombo"}}' -H "Content-Type:application/json"
```

```ballerina
service<http> hello {

    @http:resourceConfig {
        methods:["POST"],
        body:"order"
    }
    resource bindJson (http:Connection conn, http:InRequest req, json order) {
        //Access JSON field values
        json details = order.Details;
        println(details);

        json location = order.Location;
        println(location);

        http:OutResponse res = {};
        res.setJsonPayload(details);
        _ = conn.respond(res);
    }
}
```

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java [hunk 1]
```diff
@@ -130,9 +135,10 @@ public static void execute(Resource resource, BServerConnectorFuture connectorFu
                     doubleRegs[doubleParamCount++] = ((BFloat) value).floatValue();
                 } else if (btype == BTypes.typeInt) {
                     longRegs[longParamCount++] = ((BInteger) value).intValue();
-                } else if (value instanceof BStruct) {
-                    refRegs[refParamCount++] = (BRefType) value;
-                } else if (value instanceof BRefValueArray) {
+                } else if (btype == BTypes.typeBlob) {
+                    byteRegs[byteParamCount++] = ((BBlob) value).blobValue();
+                } else if (value instanceof BStruct || value instanceof BRefValueArray || value instanceof BJSON ||
+                        value instanceof BXML) {
```
Comment (anupama-pathirage) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java]: Shall we use BTypes.typeJSON and typeJSON.typeXML here?

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private List<ParamDetail> pathParams;
+    private int paramCount = 2;
+
+    public SignatureParams(HttpResource resource, List<ParamDetail> paramDetails) {
+        this.resource = resource;
+        this.paramDetails = paramDetails;
+    }
+
+    public void validate() {
+
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Unnecessary new line

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private boolean isValidResourceParam(ParamDetail paramDetail, String varTypeName) {
+        return paramDetail.getVarType().getPackagePath() != null
+                && paramDetail.getVarType().getPackagePath().equals(HttpConstants.PROTOCOL_PACKAGE_HTTP)
+                && paramDetail.getVarType().getName().equals(varTypeName);
+    }
+
+    private void validatePathParam(List<ParamDetail> paramDetails) {
+        for (int i = 0; i < paramDetails.size(); i++) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: shall we use foreach here?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/dispatching/DataBindingTest.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+import static org.ballerinalang.mime.util.Constants.CONTENT_TYPE;
+import static org.ballerinalang.mime.util.Constants.OCTET_STREAM;
+import static org.ballerinalang.mime.util.Constants.TEXT_PLAIN;
+
+/**
+ * Data binding related test cases.
+ */
+public class DataBindingTest {
+
+    CompileResult compileResult;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/dispatching/DataBindingTest.java]: Shall we make this private?

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java [hunk 5]
```diff
@@ -292,12 +288,11 @@ public static String getTextPayload(BStruct entity) {
...
         } else {
-            BStruct fileHandler = (BStruct) entity.getRefField(OVERFLOW_DATA_INDEX);
-            String filePath = fileHandler.getStringField(FILE_PATH_INDEX);
+            String filePath = getFilePathFromFileHandler(entity);
             try {
-                return new String(readFromFile(filePath), UTF_8);
+                return filePath.isEmpty() ? null : new String(readFromFile(filePath), UTF_8);
             } catch (UnsupportedEncodingException e) {
-                LOG.error("Error occured while extracting text payload from entity", e.getMessage());
+                LOG.error("Error occurred while extracting text payload from entity", e.getMessage());
```
Comment (Bhashinee) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Shall we make this variable log? 
private static final Logger log = LoggerFactory.getLogger(MimeUtil.class);

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 6]
```diff
@@ -242,6 +244,18 @@ public static void addMultipartsToCarbonMessage(HTTPCarbonMessage httpRequestMsg
         return abstractNativeFunction.getBValues(entity);
     }
 
+    /**
+     * Create file struct and set it to the entity.
+     *
+     * @param context Represent ballerina context
+     * @param entity  Represent an entity
+     */
+    private static void addFileHandlerToEntity(Context context, BStruct entity) {
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: `FileHandler` could give the wrong impression here. This simply sets a `File` struct as the `Entity` body right?

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    public SignatureParams(HttpResource resource, List<ParamDetail> paramDetails) {
+        this.resource = resource;
+        this.paramDetails = paramDetails;
+    }
+
+    public void validate() {
+
+        if (paramDetails.size() < 2) {
+            throw new BallerinaConnectorException("resource signature parameter count should be more than two");
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: `should be greater than`

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            paramCount++;
+        }
+        this.pathParams = paramDetails;
+    }
+
+    private void validateEntityBodyParam(ParamDetail entityBodyParam) {
+        String entityBodyAttributeValue = resource.getEntityBodyAttributeValue();
+        if (!entityBodyAttributeValue.equals(entityBodyParam.getVarName())) {
+            throw new BallerinaConnectorException("expected " + entityBodyAttributeValue +
+                    " as param name instead of " + entityBodyParam.getVarName());
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Shall we rephrase this to something like the following: 
``
"expected '" + entityBodyAttributeValue + "' as param name, but found '" + entityBodyParam.getVarName() + "'"
``

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java [hunk 9]
```diff
@@ -358,12 +351,16 @@ public static BXML getXmlPayload(BStruct entity) {
...
         } else {
-            BStruct fileHandler = (BStruct) entity.getRefField(OVERFLOW_DATA_INDEX);
-            String filePath = fileHandler.getStringField(FILE_PATH_INDEX);
-            return readFromFile(filePath);
+            String filePath = getFilePathFromFileHandler(entity);
+            return filePath.isEmpty() ? null : readFromFile(filePath);
         }
     }
 
+    private static String getFilePathFromFileHandler(BStruct entity) {
```
Comment (pubudu91) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Again, is `FileHandler` the correct term here, for the `File` struct?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java (hunk 1 above)
Comment (chamil321) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/connector/impl/ResourceExecutor.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java (hunk 2 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java (hunk 3 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Addressed with b98823c.

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/dispatching/DataBindingTest.java (hunk 4 above)
Comment (chamil321) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/services/dispatching/DataBindingTest.java]: Addressed with b98823c.

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java (hunk 5 above)
Comment (chamil321) [reply stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java (hunk 6 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java (hunk 7 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java (hunk 8 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Addressed with b98823c.

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java (hunk 9 above)
Comment (chamil321) [reply stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/MimeUtil.java]: Addressed with b98823c.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java [hunk 10]
```diff
@@ -199,8 +207,58 @@ public static HttpResource findResource(HTTPServicesRegistry servicesRegistry,
...
+        }
+
+        if (signatureParams.getEntityBody() == null) {
+            return bValues;
+        }
+        try {
+            bValues[bValues.length - 1] = populateAndGetEntityBody(httpResource, inRequest, inRequestEntity,
+                    signatureParams.getEntityBody().getVarType());
+        } catch (BallerinaException ex) {
+            httpCarbonMessage.setProperty(HttpConstants.HTTP_STATUS_CODE, 400);
```
Comment (Manuri) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Better to have a constant which reflects the meaning of the HTTP status code 400 here, for better readability and maintainability.

Comment (VijithaEkanayake) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Can't We use HttpURLConnection.HTTP_BAD_REQUEST instead of 400?

Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Ack

Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpDispatcher.java]: Ack

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * @since 0.963.0
+ */
+public class SignatureParams {
+
+    private HttpResource resource;
+    private final List<ParamDetail> paramDetails;
+    private ParamDetail entityBody;
+    private List<ParamDetail> pathParams;
+    private int paramCount = 2;
```
Comment (shafreenAnfar) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: shall we add a comment explaining the "2"

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            validatePathParam(paramDetails.subList(2, paramDetails.size()));
+        } else {
+            int lastParamIndex = paramDetails.size() - 1;
+            validatePathParam(paramDetails.subList(2, lastParamIndex));
+            validateEntityBodyParam(paramDetails.get(lastParamIndex));
+        }
+    }
+
+    private boolean isValidResourceParam(ParamDetail paramDetail, String varTypeName) {
+        return paramDetail.getVarType().getPackagePath() != null
```
Comment (shafreenAnfar) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: shall avoid repeating..

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java [hunk 13]
```diff
@@ -44,22 +47,47 @@ public BallerinaHTTPConnectorListener(HTTPServicesRegistry httpServicesRegistry)
...
-        HttpResource httpResource = HttpDispatcher.findResource(httpServicesRegistry, httpCarbonMessage);
+        HttpResource httpResource;
+        if (!isAccessed(httpCarbonMessage)) {
+            httpResource = HttpDispatcher.findResource(httpServicesRegistry, httpCarbonMessage);
+            if (HttpDispatcher.isDiffered(httpResource)) {
+                httpCarbonMessage.setProperty(HTTP_RESOURCE, httpResource);
+            } else {
+                extractPropertiesAndStartResourceExecution(httpCarbonMessage, httpResource);
+            }
+        } else {
```
Comment (lankavitharana) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java]: shall we remove the else part?

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java [hunk 14]
```diff
@@ -44,22 +47,47 @@ public BallerinaHTTPConnectorListener(HTTPServicesRegistry httpServicesRegistry)
...
+            httpResource = (HttpResource) httpCarbonMessage.getProperty(HTTP_RESOURCE);
+            extractPropertiesAndStartResourceExecution(httpCarbonMessage, httpResource);
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        log.error("Error in http server connector", throwable);
```
Comment (lankavitharana) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java]: log error msg as well?
log.error("Error in http server connector" + throwable.getMessage(), throwable);

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java [hunk 15]
```diff
@@ -44,22 +47,47 @@ public BallerinaHTTPConnectorListener(HTTPServicesRegistry httpServicesRegistry)
...
-        Map<String, Object> properties = null;
         if (httpCarbonMessage.getProperty(HttpConstants.SRC_HANDLER) != null) {
             Object srcHandler = httpCarbonMessage.getProperty(HttpConstants.SRC_HANDLER);
             properties = Collections.singletonMap(HttpConstants.SRC_HANDLER, srcHandler);
         }
-        BValue[] signatureParams = HttpDispatcher.getSignatureParameters(httpResource, httpCarbonMessage);
+        BValue[] signatureParams;
+        try {
+            signatureParams = HttpDispatcher.getSignatureParameters(httpResource, httpCarbonMessage);
+        } catch (BallerinaException ex) {
```
Comment (lankavitharana) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java]: I feel like we shouldn't handle this here

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java (hunk 14 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/BallerinaHTTPConnectorListener.java]: Ack

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java (hunk 12 above)
Comment (chamil321) [reply stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/SignatureParams.java]: Ack
